OpenSim/Simulation/Model/ScapulothoracicJoint.cpp

```diff
@@ -0,0 +1,288 @@
+{
+    constructProperties();
+    upd_thoracic_ellipsoid_radii_x_y_z() = ellipsoidRadii;
+    upd_scapula_winging_axis_origin(0) = wingingOrigin[0];
+    upd_scapula_winging_axis_origin(1) = wingingOrigin[1];
+    upd_scapula_winging_axis_direction() = wingingDirection;
+}
+
+//=============================================================================
+// CONSTRUCTION
+//=============================================================================
+//_____________________________________________________________________________
+/**
+ * Construct properties with their default values.
+ */
+void ScapulothoracicJoint::constructProperties()
+{
+    setAuthors("Ajay Seth");
+    constructProperty_thoracic_ellipsoid_radii_x_y_z(Vec3(NaN));
+    constructProperty_scapula_winging_axis_origin(Vector(2, 0.0));
+    constructProperty_scapula_winging_axis_direction(0.0);
+}
+
+
+//=============================================================================
+// SCALING
+//=============================================================================
+//_____________________________________________________________________________
+/**
+* Scale a joint based on XYZ scale factors for the bodies.
+*
+* @param aScaleSet Set of XYZ scale factors for the bodies.
+* @todo Need to scale transforms appropriately, given an arbitrary axis.
+*/
+void ScapulothoracicJoint::extendScale(const SimTK::State& s,
+                                       const ScaleSet& scaleSet)
+{
+    // Joint knows how to scale locations of the joint in parent and on the body
+    Super::extendScale(s, scaleSet);
+
+    const std::string& parentName = getParentFrame().getName();
+
+    // Scaling related to the parent body:
+    //
+    // Joint kinematics are scaled by the scale factors for the parent body,
+    // so get those body's factors
+    Vec3 scaleFactors(1.0);
+    for (int i = 0; i < scaleSet.getSize(); i++) {
+        const Scale &scale = scaleSet.get(i);
+
+        if (scale.getSegmentName() == parentName) {
+            scale.getScaleFactors(scaleFactors);
+            break;
+        }
+    }
+
+    // Scale the size of the mobilizer
+    for (int i=0; i<3; i++) {
+        upd_thoracic_ellipsoid_radii_x_y_z()[i] *= scaleFactors[i];
+    }
+}
+
+//=============================================================================
+// Simbody Model building.
+//=============================================================================
+//_____________________________________________________________________________
+void ScapulothoracicJoint::extendAddToSystem(SimTK::MultibodySystem& system) const
+{
+    Super::extendAddToSystem(system);
+
+    // Transform for ellipsoid joint frame in intermediate Scapula massless
+    // body frame.
+    //
+    // Oriented such that intermediate frame is aligned with scapula joint
+    // frame with respect to the scapula body frame as user specified by
+    // _location and _orientation
+    Transform ellipsoidJointFrameInIntermediate{
```

Comment:
Generally ok within classes, for public interfaces we need to be careful since creation of objects at the call sites or their memory management could be problematic for scripting users.

Reply:
Just so I understand, what can happen with scripting users? Do interface generators (e.g. SWIG) not obey value semantics, such that writing (e.g. in Java):

```java
SomeClass.f(new SomeValue());
```

Can bug out? What I mean is, what is there to be "careful" about here for public interfaces? Exposing C++ APIs that take values? Exposing C++ APIs that take rvalues? Writing scripts in the bound-to language (e.g. Matlab/Python) that create temporaries?

When writing C++, I'm mostly just writing standard modern C++, so there's a risk that I'm not aware of this nuance, and that I might inject bugs into downstream (e.g. the GUI) without the tests (written in C++) catching them.